### DISCUSSION TITLE
The game freezes while searching the database

### DISCUSSION THREAD

**Discussion by 10100010010001001010** [axmolengine/axmol#1616] [Q&A] [answered]
How to implement the program working while searching the database? It's supposed to be a very long search, but during it the application freezes and even crashed a couple of times if you try to press something. Use multithreading?

**Comment by rh101** [comment, marked as answer]
> Use multithreading?

Exactly that. When the worker thread is ready to send data back to the main thread, then use the `Scheduler::runOnAxmolThread()` for that.

**Reply by 10100010010001001010** [reply]
Is it possible get the program to run fine without creating threads? What I mean is that the program won't crash if I try to press something. At the moment, if I try to interact with it while searching, it freezes and most likely crashes.

**Comment by rh101** [comment]
> Is it possible get the program to run fine without creating threads? What I mean is that the program won't crash if I try to press something. At the moment, if I try to interact with it while searching, it freezes and most likely crashes.

If it's "freezing" because it is performing some task that requires waiting for a result (I/O, DB access, etc.), then you really can't get around the fact that you need to do that in a separate thread if you want the UI (on main thread) to still be responsive.

As for why it's crashing, I don't know; are you certain there isn't a bug in your code?

**Reply by 10100010010001001010** [reply]
As for the crashes, they happened when I tried to interact with a frozen application. I corrected the errors and this has not happened yet. That is, in theory, the application should respond even if it endlessly reads the database, despite the fact that it will not require I/O from the user?

**Reply by rh101** [reply]
Well, not quite.  

If your database access routine is in the main thread, which also happens to be the thread updating the UI/input etc., then it can only process things synchronously, meaning it must wait for each call to finish before continuing to the next. 

For example:

```
void Update()
{
    auto data = GetDataFromDB(); // It will need to wait till this method returns before continuing, meaning it'll "freeze" for as long as required for it to finish.  The UI will not be responsive.
    DoSomethingWithData(data);
    UpdateUI();
    GetInput();
}
```
